spock-specs/src/test/groovy/org/spockframework/smoke/extension/TempDirExtensionSpec.groovy

```diff
@@ -0,0 +1,104 @@
+package org.spockframework.smoke.extension
+
+import org.spockframework.util.Beta
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.lang.Stepwise
+import spock.lang.TempDir
+
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.nio.file.attribute.PosixFilePermissions
+
+/**
+ * @author dqyuan
+ */
+@Beta
+@Stepwise
+class TempDirExtensionSpec extends Specification {
+
+  @Shared
+  @TempDir
+  Path sharedDir
+  static Path preShared
+
+  @TempDir
+  File iterationDir
+  static File preIteration
+
+  @TempDir(baseDir = "build")
+  Path customerTempDir
+
+  @TempDir(baseDir = "build", reserveAfterTest = { true })
+  Path reserveDir
+  static Path reserveFile
+
+  def "temp dir exist"() {
+    preShared = sharedDir
+    preIteration = iterationDir
+    reserveFile = reserveDir.resolve("test.txt")
+    Files.write(reserveFile, "aaa".getBytes())
+    expect:
+    assert Files.exists(sharedDir)
+    assert Files.exists(Paths.get(iterationDir.toURI()))
+    assert Files.exists(customerTempDir)
+    assert Files.exists(reserveDir)
+    assert customerTempDir.parent.toString().startsWith("build")
+    assert reserveDir.parent.toString().startsWith("build")
+  }
+
+  def "reserve temp directory when reserveAfterTest is true"() {
+    expect:
+    assert Files.exists(reserveFile)
+  }
+
+  def "@TempDir creates only one dir for one spec, if the annotated field is 'shared'"() {
+    expect:
+    assert sharedDir == preShared
+    assert iterationDir != preIteration
+    assert !Files.exists(Paths.get(preIteration.toURI()))
+  }
+
+  def "@TempDir creates one temp directory per iteration for normal field"(String foo, int bar) {
+    expect:
+    assert iterationDir != preIteration
+
+    when:
+    preIteration = iterationDir
```

Comment:
Maybe better do this in a `cleanup:` block.
It also does not really belong there semantically, but I think better than in a `when:` block